libs/ts-rest/next/src/lib/ts-rest-next.ts

```diff
@@ -186,42 +197,96 @@ export const createNextRoute = <T extends AppRouter>(
 export const createNextRouter = <T extends AppRouter>(
   routes: T,
   obj: RecursiveRouterObj<T>,
-  options?: {
-    jsonQuery?: boolean;
-    responseValidation?: boolean;
-    throwRequestValidation?: boolean;
-    errorHandler?: (
-      err: unknown,
-      req: NextApiRequest,
-      res: NextApiResponse,
-    ) => void;
-  },
+  options?: CreateNextRouterOptions,
 ) => {
-  const {
-    jsonQuery = false,
-    responseValidation = false,
-    throwRequestValidation = false,
-  } = options || {};
+  return handlerFactory(function (req) {
+    const combinedRouter = mergeRouterAndImplementation(routes, obj);
 
-  const combinedRouter = mergeRouterAndImplementation(routes, obj);
-
-  return async (req: NextApiRequest, res: NextApiResponse) => {
+    // eslint-disable-next-line prefer-const
     let { 'ts-rest': params, ...query } = req.query;
-    params = (params as string[]) || [];
 
+    params = (params as string[]) || [];
     const route = getRouteImplementation(
       combinedRouter,
       params,
       req.method as string,
     );
+    let pathParams;
+    if (route) {
+      pathParams = getPathParamsFromArray(params, route);
+    } else {
+      pathParams = {};
+    }
+    return { pathParams, query, route };
+  }, options);
+};
 
+/**
+ * Turn a contract route and a handler into a Next.js compatible handler
+ * Should be exported from your pages/api/path/to/handler.tsx file.
+ *
+ * e.g.
+ *
+ * ```typescript
+ * export default createNextRouter(contract, implementationHandler);
+ * ```
+ *
+ * @param appRoute
+ * @param options
+ * @returns
+ */
+export function createSingleRouteHandler<T extends AppRoute>(
+  appRoute: AppRoute,
+  implementationHandler: AppRouteImplementation<T>,
+  options?: CreateNextRouterOptions,
+) {
+  return handlerFactory((req) => {
+    const route = { ...appRoute, implementation: implementationHandler };
+    const pathParams = req.query as Record<string, string>;
+    const query = req.query;
+
+    const isValidRoute = isRouteCorrect(
+      appRoute,
+      req.url!.split('/').slice(1),
+      req.method as string,
+    );
```

Comment:
Although the current code should work in most cases. The passed `pathParams` will be wrong if the route file/dir name `[foo]` does not match the name in the contract, despite `isValidRoute` being true due to using the actual URL rather than `req.query`. 

I think it'll also be better to also filter out the `pathParams` from `req.query` before passing it into `query`, but it won't really work if the names in the original `req.query` and `pathParams` don't match, but there's nothing we could really do at this point.

```suggestion
    const urlChunks = req.url!.split('/').slice(1);
    const pathParams = getPathParamsFromArray(urlChunks, route);
    const query = req.query
      ? Object.fromEntries(
          Object.entries(req.query).filter(
            ([key]) => pathParams[key] === undefined,
          ),
        )
      : {};

    const isValidRoute = isRouteCorrect(
      appRoute,
      urlChunks,
      req.method as string,
    );
```